late/chroot_scripts/60-detect-no-recovery-patition.py

```diff
@@ -0,0 +1,115 @@
+import os
+import shutil
+import subprocess as cmd
+
+devices = []
+
+for name in glob.glob('/sys/block/*'):
+    name = basename(name)
+    if name.startswith('sd'):
+        devices.append(name)
+    elif name.startswith('md'):
+        devices.append(name)
+    elif name.startswith('nvme'):
+        devices.append(name)
+    elif name.startswith('pmem'):
+        devices.append(name)
+
+root = cmd.check_output("findmnt -M / -o source -v | tail -n1", shell=True)
+root = basename(root.decode().strip())
+root_device = ''
+
+cdrom = cmd.check_output("findmnt -M /cdrom -o source -v | tail -n1",
+                         shell=True)
+cdrom = basename(cdrom.decode().strip())
+cdrom_device = ''
+
+disks = json.loads(cmd.check_output("lsblk -fs -J", shell=True).decode())
+disks = disks['blockdevices']
+
+for disk in disks:
+    if disk['name'] == root:
+        data = json.dumps(disk)
+        for device in devices:
+            if device in data:
+                root_device = device
+                break
+    elif disk['name'] == cdrom:
+        data = json.dumps(disk)
+        for device in devices:
+            if device in data:
+                cdrom_device = device
+                break
+
+if root_device == cdrom_device:
+    exit(0)
+
+print('No recovery partition is detected.')
+
+
+def check_depends(pkg_name, depends=None):
+    if depends is None:
+        depends = []
+
+    if pkg_name not in cache:
+        return depends
+
+    pkg = cache[pkg_name]
+
+    if not pkg.has_versions:
+        return depends
+
+    depends_list = pkg.version_list[0].depends_list
+
+    if not depends_list:
+        return depends
+
+    if 'Depends' in depends_list:
+        for dep in depends_list['Depends']:
+            pkg_name = dep[0].all_targets()[0].parent_pkg.name
+            if pkg_name not in depends and not cache[pkg_name].current_ver:
+                depends.append(pkg_name)
+
+    return depends
+
+
+cache = apt.apt_pkg.Cache()
+langs = ("ca", "cs", "da", "de", "en", "en_US", "es", "eu", "fr", "gl", "it",
+         "hu", "nl", "pl", "pt", "pt_BR", "sl", "fi", "sv", "el", "bg", "ru",
+         "ko", "zh-hans", "zh-hant", "ja")
+depends = []
+
+for lang in langs:
+    pkgs = cmd.check_output('check-language-support'
+                            + ' --show-installed'
+                            + ' -l ' + lang,
+                            shell=True)
+    pkgs = pkgs.decode('utf-8').strip().split(' ')
+    for pkg in pkgs:
+        if pkg not in depends and not cache[pkg].current_ver:
+            depends.append(pkg)
+
+pre_len = len(depends)
+
+while True:
+    for dep in depends.copy():
+        check_depends(dep, depends)
+    if len(depends) != pre_len:
+        pre_len = len(depends)
+    else:
+        break
+
+os.makedirs('/dell/debs')
+
+for folder in ('/cdrom/pool', '/cdrom/debs'):
+    for deb in glob.glob(f'{folder}/**/*.deb', recursive=True):
+        pkg_name = cmd.check_output(f'dpkg-deb -f {deb} Package', shell=True)
```

Comment:
Is the path not available in the apt cache?  I would think that the performance would be better querying from that rather than checking every single debian package.  If it turns out it's not available, I guess this works though.

Reply:
It is clearer to check all physical Debian packages directly. There are too many packages in apt cache to check if they are available in the local archive.

Reply:
I did not find the full path of the Debian packages yet so I use `apt-get download` instead.
Is it OK?